components/concept/storage/src/main/java/mozilla/components/concept/storage/BookmarksStorage.kt

```diff
@@ -0,0 +1,82 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.concept.storage
+
+import mozilla.appservices.places.BookmarkFolder
+import mozilla.appservices.places.BookmarkItem
+import mozilla.appservices.places.BookmarkRoot
+import mozilla.appservices.places.BookmarkTreeNode
+import mozilla.appservices.places.BookmarkUpdateInfo
+
+interface BookmarksStorage {
+
+    /**
+     * Produces a populated bookmarks tree for the given root.
+     *
+     * @param root The bookmark root to obtain.
+     * @param recursive Whether to recurse and obtain all levels of children.
+     * @return The populated root.
+     */
+    suspend fun getTree(root: BookmarkRoot, recursive: Boolean = false): BookmarkTreeNode?
+
+    /**
+     * Produces a populated bookmarks tree for the given guid string.
+     *
+     * @param guid The bookmark guid to obtain.
+     * @param recursive Whether to recurse and obtain all levels of children.
+     * @return The populated root starting from the guid.
+     */
+    suspend fun getTree(guid: String, recursive: Boolean = false): BookmarkTreeNode?
+
+    /**
+     * Searches bookmarks with a query string.
+     *
+     * @param query The query string to search.
+     * @param limit The maximum number of entries to return.
+     * @return The list of matching BookmarkItems.
+     */
+    suspend fun searchBookmarks(query: String, limit: Int = DEFAULT_BOOKMARK_SEARCH_LIMIT): List<BookmarkItem>
+
+    /**
+     * Adds a new bookmark tree node to a given node.
+     *
+     * @param newNode The node to add.
+     * @param position The optional position to add it in.
+     * @return The guid of the newly inserted bookmark item.
+     */
+    suspend fun addNode(target: BookmarkFolder, newNode: BookmarkTreeNode, position: Int? = null): String
```

Comment:
s/target/parent. Also, consider if you just want to use `parentGuid` here. I think it'll be much easier down the road from the consumption point of view.

regarding `newNode` - I don't think this will map well to how places operates. `BookmarkTreeNode` requires a `guid`, timestamps, and some other things - it's not application's job to generate those bits of information. And so, that object  type shouldn't be used for creation.

Additionally, the places API specifically provides us with three `create` endpoints, partially to remove potential footguns. I think we should just expose these three `create` methods at the concept level.

Reply:
Done